### ISSUE TITLE
qbittorrent when using virtual machine

### ISSUE THREAD

**Issue by yekul** [mdhiggins/sickbeard_mp4_automator#1319]
Apologies for using this like a help desk but having quite an issue making this work and can't find other similar searches. Automator is working flawlessly for sabdnzbd/sonarr btw so thank you very much. 

Current setup is Sonarr/sabnzbd/sickbeard mp4 automator on native windows, then qbittorrent loaded up on a virtual machine on that. They can communicate no problems if you take automator out of the equation (no import issues etc), however I am having a difficult time trying to get the script aspect to work. I am trying to get it to run the process script on the Sonarr machine as that has more resources (the VM only has low cores+ram), but I simply can't figure out how to make this happen. 

Or is this simply not a viable option?

**Comment by mdhiggins** [comment]
Setting up postSonarr.py should take care of everything, the downloader scripts are optional (qBittorrent). That being said however you'll need to use completed download handling for that to work otherwise the downloads will never import.

There's no good option to use the script with SAB and not with qBittorrent though (because having completed download handling will interfere with the manual conversion of SAB).

You can just set up qBittorrent's script and turn off conversion though

Either way you should have postSonarr.py setup

In summary, your 2 options are

- postSonarr.py
- completed download handling enabled in Soanrr
- no downloader scripts (no SABPostProcess.py and no qBittorrentPostProcess.py)

OR

- postSonarr.py
- completed download handling disabled in Sonarr
- SABPostProcess.py setup in SAB
- qBittorrentPostProcess.py setup in qBittorrent but with `convert = False` in [qBittorrent] in autoProcess.ini

**Comment by yekul** [comment]
I think i'm a complete moron. I had setup postSonarr.py (added the connection etc) assuming that it was required ON TOP of the relevant script (sabpostprocess.py) and disabled completed download handling. So basically up till this point it has been using sabpostprocess.py and the postSonarr.py was completely unnecessary right? 

So am I right in thinking the major downside to using postSonarr.py is that it will do the conversion after it has been moved to the final folder as specified in Sonarr? Fingers crossed I now understand, as that would work perfectly for me and I can ignore the individual scripts all together. If that's right, so sorry for wasting your time...

**Comment by mdhiggins** [comment]
That is all correct

The plus side of using the extra scripts is you get your conversion before moving

postSonarr will do the conversion if it hasn't happened. It also does other things that the SAB/qBT scripts don't do (tagging/qtfs) so you always want that script to be installed but the others are optional. Just turn on completed download handling if you're not going to use the SAB/qBT scripts and you should be set

**Comment by yekul** [comment]
Man oh man feel so stupid right now. Thank you so much for your help it's greatly appreciated.

**Comment by yekul** [comment]
Apologies, one last thing. Since moving to this setup I am running into one glaring issue that i'm sure is fixable. Because it imports the file first, the naming convention ends up being wrong for me. I have the codec listed as part of the title, but as i'm encoding from x264 to x265 I need the file to be renamed again after the postSonarr.py takes place as otherwise Sonarr assumes the file is x264 as that's what it was during initial import; is that possible?

**Comment by mdhiggins** [comment]
The final step of postSonarr is that it should trigger a renameSeries via the API. If that's not working double check your logs and autoProcess.ini settings cause without the correct api data that won't be successful

You should see something in this format
"Sonarr response RenameSeries command: ID %d %s."

**Comment by yekul** [comment]
Hrm curious. So yes, it is appearing to do that as at the end of the processing by the script I get: 
2020-09-07 23:04:39 - SonarrPostProcess - INFO - Sonarr response RenameSeries command: ID 614028 queued. 

But, for whatever reason it doesn't seem to acknowledge that the file has changed. As in, on Sonarr's end even if I go in and manually click the rename files it will say 'Success, no more files to rename!'. However when I was previously using the Sabnzbd script there would be no issues in it recognising the file changes and renaming during import. 

This feels like it could be a Sonarr issue (which I truly don't know how to address, but makes it very much not your problem)? However i'm also wondering if there's some difference between postsonarr.py and SABpostprocess.py that may not be updating the codec information or something? If I inspect the files with mediainfo they are showing correctly as HEVC files so I really am not sure and leaning towards most likely a Sonarr issue.

**Comment by yekul** [comment]
Hrm perhaps this is still coming into play: 
https://github.com/Sonarr/Sonarr/issues/3366

So even though a rescan is forced, because the file name was the same it now just says 'Yep that's all good, not changing a thing'. Which fits, I literally removed a '-' from the file name, did a refresh/rescan and then the file renaming worked perfectly fine. I have no idea how to get round this other than perhaps having the script add a -1 to the filename or similar so it will force a rename to actually occur.

**Comment by mdhiggins** [comment]
Hm that's annoying

See if this modified version of postSonarr.py fixes that for you
```
#!/usr/bin/env python3
import os
import sys
import requests
import time
from resources.log import getLogger
from resources.readsettings import ReadSettings
from resources.metadata import MediaType
from resources.mediaprocessor import MediaProcessor


def rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log, retries=6, delay=10):
    headers = {'X-Api-Key': apikey}
    # First trigger rescan
    payload = {'name': 'RescanSeries', 'seriesId': seriesid}
    url = protocol + host + ":" + str(port) + webroot + "/api/command"
    r = requests.post(url, json=payload, headers=headers)
    rstate = r.json()
    try:
        rstate = rstate[0]
    except:
        pass
    log.info("Sonarr response RescanSeries command: ID %d %s." % (rstate['id'], rstate['state']))
    log.debug(str(rstate))

    # Then wait for it to finish
    url = protocol + host + ":" + str(port) + webroot + "/api/command/" + str(rstate['id'])
    log.info("Requesting episode information from Sonarr for series ID %s." % seriesid)
    r = requests.get(url, headers=headers)
    command = r.json()
    attempts = 0
    while command['state'].lower() not in ['complete', 'completed'] and attempts < retries:
        log.info("State: %s." % (command['state']))
        time.sleep(delay)
        r = requests.get(url, headers=headers)
        command = r.json()
        attempts += 1
    log.info("Final state: %s." % (command['state']))
    log.debug(str(command))
    return command['state'].lower() in ['complete', 'completed']


def getEpisodeInformation(host, port, webroot, apikey, protocol, episodeid, log):
    headers = {'X-Api-Key': apikey}
    url = protocol + host + ":" + str(port) + webroot + "/api/episode?seriesId=" + seriesid
    log.info("Requesting updated episode information from Sonarr for series ID %s." % seriesid)
    r = requests.get(url, headers=headers)
    payload = r.json()
    sonarrepinfo = None
    for ep in payload:
        if int(ep['episodeNumber']) == episode and int(ep['seasonNumber']) == season:
            return ep
    return None


def renameFile(inputfile, log):
    filename, fileext = os.path.splitext(inputfile)
    outputfile = "%s.rnm%s" % (filename, fileext)
    i = 2
    while os.path.isfile(outputfile):
        outputfile = "%s.rnm%d%s" % (filename, i, fileext)
        i += 1
    os.rename(inputfile, outputfile)
    log.debug("Renaming file %s to %s." % (inputfile, outputfile))
    return outputfile


def renameSeries(host, port, webroot, apikey, protocol, seriesid, log):
    headers = {'X-Api-Key': apikey}
    # First trigger rescan
    payload = {'name': 'RenameSeries', 'seriesIds': [seriesid]}
    url = protocol + host + ":" + str(port) + webroot + "/api/command"
    r = requests.post(url, json=payload, headers=headers)
    rstate = r.json()
    try:
        rstate = rstate[0]
    except:
        pass
    log.info("Sonarr response RenameSeries command: ID %d %s." % (rstate['id'], rstate['state']))
    log.debug(str(rstate))


log = getLogger("SonarrPostProcess")

log.info("Sonarr extra script post processing started.")

if os.environ.get('sonarr_eventtype') == "Test":
    sys.exit(0)

settings = ReadSettings()

log.debug(os.environ)

inputfile = os.environ.get('sonarr_episodefile_path')
original = os.environ.get('sonarr_episodefile_scenename')
tvdb_id = int(os.environ.get('sonarr_series_tvdbid'))
season = int(os.environ.get('sonarr_episodefile_seasonnumber'))
seriesid = os.environ.get('sonarr_series_id')

try:
    episode = int(os.environ.get('sonarr_episodefile_episodenumbers'))
except:
    episode = int(os.environ.get('sonarr_episodefile_episodenumbers').split(",")[0])

mp = MediaProcessor(settings)

log.debug("Input file: %s." % inputfile)
log.debug("Original name: %s." % original)
log.debug("TVDB ID: %s." % tvdb_id)
log.debug("Season: %s episode: %s." % (season, episode))
log.debug("Sonarr series ID: %s." % seriesid)

try:
    try:
        inputfile = renameFile(inputfile, log)
    except:
        log.exception("Error renaming inputfile")

    success = mp.fullprocess(inputfile, MediaType.TV, tvdbid=tvdb_id, season=season, episode=episode, original=original)

    if success:
        # Update Sonarr to continue monitored status
        try:
            host = settings.Sonarr['host']
            port = settings.Sonarr['port']
            webroot = settings.Sonarr['webroot']
            apikey = settings.Sonarr['apikey']
            ssl = settings.Sonarr['ssl']
            protocol = "https://" if ssl else "http://"

            log.debug("Sonarr host: %s." % host)
            log.debug("Sonarr port: %s." % port)
            log.debug("Sonarr webroot: %s." % webroot)
            log.debug("Sonarr apikey: %s." % apikey)
            log.debug("Sonarr protocol: %s." % protocol)

            if apikey != '':
                headers = {'X-Api-Key': apikey}

                if rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log):
                    log.info("Rescan command completed")

                    sonarrepinfo = getEpisodeInformation(host, port, webroot, apikey, protocol, seriesid, log)
                    if not sonarrepinfo:
                        log.error("No valid episode information found, aborting.")
                        sys.exit(1)

                    if not sonarrepinfo.get('hasFile'):
                        log.warning("Rescanned episode does not have a file, attempting second rescan.")
                        if rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log):
                            sonarrepinfo = getEpisodeInformation(host, port, webroot, apikey, protocol, seriesid, log)
                            if not sonarrepinfo:
                                log.error("No valid episode information found, aborting.")
                                sys.exit(1)
                            if not sonarrepinfo.get('hasFile'):
                                log.warning("Rescanned episode still does not have a file, will not set to monitored to prevent endless loop.")
                                sys.exit(1)
                            else:
                                log.info("File found after second rescan.")
                        else:
                            log.error("Rescan command timed out")
                            sys.exit(1)

                    # Then set that episode to monitored
                    sonarrepinfo['monitored'] = True
                    log.debug("Sending PUT request with following payload:")
                    log.debug(str(sonarrepinfo))

                    url = protocol + host + ":" + str(port) + webroot + "/api/episode/" + str(sonarrepinfo['id'])
                    r = requests.put(url, json=sonarrepinfo, headers=headers)
                    success = r.json()

                    log.debug("PUT request returned:")
                    log.debug(str(success))
                    log.info("Sonarr monitoring information updated for episode %s." % success['title'])

                    renameSeries(host, port, webroot, apikey, protocol, seriesid, log)
                else:
                    log.error("Rescan command timed out")
                    sys.exit(1)
            else:
                log.error("Your Sonarr API Key is blank. Update autoProcess.ini to enable status updates.")
        except:
            log.exception("Sonarr monitor status update failed.")
    else:
        log.info("Processing returned False.")
        sys.exit(1)
except:
    log.exception("Error processing file")
    sys.exit(1)
```

**Comment by yekul** [comment]
Amazing. Worked perfectly. Renaming and all. Unsure what you changed but will have a sift through to compare files when I get a chance and make sure if I do any future updates I can resolve this myself with whatever code snippets need to be modified. Fingers crossed Sonarr eventually implements the recommendations and this will be a non-issue. 

Thank you so very much. Your help has been greatly appreciated.

**Comment by mdhiggins** [comment]
6341b56163ffffcf6e5864bbd14d2596c1adbb2e

I merged it into the master branch

I basically just renamed the file to filename.rnm.ext before any processing takes place which is enough to trick sonarr/radarr that the rename needs to be recalculated at the end

Try that update and make sure its all working. I applied it to radarr as well

**Comment by yekul** [comment]
> Try that update and make sure its all working. I applied it to radarr as well

Just ran through a test file and working flawlessly thanks again :D

**Comment by joebeem** [comment]
I happened to find this issue while investigating why newer files from sonarr/radarr are being imported with a ".rnm" appended to the file. Considering no one else seems to be speaking about such an issue leads me to believe it's something with my settings. The settings have worked as configured for the past 3 years, but maybe I am not using optimal settings anymore..

**Comment by mdhiggins** [comment]
The .rnm component should be temporary, at the end of processing sonarr/radarr should run a rename sweep and fix that, is that not happening?

**Comment by joebeem** [comment]
Thank you for responding. I am actually finding even more recently that I'll have multiple iterations of the same file from both radarr and sonarr where file name will be appended with .rnm, .rnm2, .rnm3, etc. I use this project together with sabnzbd and it will continue to download the same file continuously. This only recently started happening. I'll double check the settings, but I wonder if I reverting to an earlier version in git would temporary fix as from what I can tell, this is isolated to me.

**Comment by mdhiggins** [comment]
Can you post some logs so I can see if your API calls are failing for some reason? And maybe some logs from Sonarr/Radarr too?

**Comment by mdhiggins** [comment]
Reverting to a previous version would fix it but this is likely a symptom of some other problem and perhaps can be fixed, but if its causing issues like this I'd rather address it for future users cause it probably won't be just you that's impacted long term

**Comment by joebeem** [comment]
Sure, I would be happy to post logs from all the apps. I will try posting them today.

**Comment by mdhiggins** [comment]
The SonarrPostProcess components and stuff about API calls from Sonarr should be most helpful
You can turn on debug level logging too by setting that in logging.ini

**Comment by joebeem** [comment]
I don't post on github too much, so I am not sure about best practices when it comes to posting logs. I will try to just attach to the post. Is this helpful at all?

[sonarr-debug.log](https://github.com/mdhiggins/sickbeard_mp4_automator/files/5253523/sonarr-debug.log)

**Comment by mdhiggins** [comment]
See if this version of postSonarr fixes the issue

```
#!/usr/bin/env python3
import os
import sys
import requests
import time
from resources.log import getLogger
from resources.readsettings import ReadSettings
from resources.metadata import MediaType
from resources.mediaprocessor import MediaProcessor


def rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log, retries=6, delay=10):
    headers = {'X-Api-Key': apikey}
    # First trigger rescan
    payload = {'name': 'RescanSeries', 'seriesId': seriesid}
    url = protocol + host + ":" + str(port) + webroot + "/api/command"
    r = requests.post(url, json=payload, headers=headers)
    rstate = r.json()
    try:
        rstate = rstate[0]
    except:
        pass
    log.info("Sonarr response RescanSeries command: ID %d %s." % (rstate['id'], rstate['state']))
    log.debug(str(rstate))

    # Then wait for it to finish
    url = protocol + host + ":" + str(port) + webroot + "/api/command/" + str(rstate['id'])
    log.info("Requesting episode information from Sonarr for series ID %s." % seriesid)
    r = requests.get(url, headers=headers)
    command = r.json()
    attempts = 0
    while command['state'].lower() not in ['complete', 'completed'] and attempts < retries:
        log.info("State: %s." % (command['state']))
        time.sleep(delay)
        r = requests.get(url, headers=headers)
        command = r.json()
        attempts += 1
    log.info("Final state: %s." % (command['state']))
    log.debug(str(command))
    return command['state'].lower() in ['complete', 'completed']


def getEpisodeInformation(host, port, webroot, apikey, protocol, episodeid, log):
    headers = {'X-Api-Key': apikey}
    url = protocol + host + ":" + str(port) + webroot + "/api/episode?seriesId=" + seriesid
    log.info("Requesting updated episode information from Sonarr for series ID %s." % seriesid)
    r = requests.get(url, headers=headers)
    payload = r.json()
    sonarrepinfo = None
    for ep in payload:
        if int(ep['episodeNumber']) == episode and int(ep['seasonNumber']) == season:
            return ep
    return None


def renameFile(inputfile, log):
    filename, fileext = os.path.splitext(inputfile)
    outputfile = "%s.rnm%s" % (filename, fileext)
    i = 2
    while os.path.isfile(outputfile):
        outputfile = "%s.rnm%d%s" % (filename, i, fileext)
        i += 1
    os.rename(inputfile, outputfile)
    log.debug("Renaming file %s to %s." % (inputfile, outputfile))
    return outputfile


def renameSeries(host, port, webroot, apikey, protocol, seriesid, log):
    headers = {'X-Api-Key': apikey}
    # First trigger rescan
    payload = {'name': 'RenameSeries', 'seriesIds': [seriesid]}
    url = protocol + host + ":" + str(port) + webroot + "/api/command"
    r = requests.post(url, json=payload, headers=headers)
    rstate = r.json()
    try:
        rstate = rstate[0]
    except:
        pass
    log.info("Sonarr response RenameSeries command: ID %d %s." % (rstate['id'], rstate['state']))
    log.debug(str(rstate))


log = getLogger("SonarrPostProcess")

log.info("Sonarr extra script post processing started.")

if os.environ.get('sonarr_eventtype') == "Test":
    sys.exit(0)

settings = ReadSettings()

log.debug(os.environ)

inputfile = os.environ.get('sonarr_episodefile_path')
original = os.environ.get('sonarr_episodefile_scenename')
tvdb_id = int(os.environ.get('sonarr_series_tvdbid'))
season = int(os.environ.get('sonarr_episodefile_seasonnumber'))
seriesid = os.environ.get('sonarr_series_id')

try:
    episode = int(os.environ.get('sonarr_episodefile_episodenumbers'))
except:
    episode = int(os.environ.get('sonarr_episodefile_episodenumbers').split(",")[0])

mp = MediaProcessor(settings)

log.debug("Input file: %s." % inputfile)
log.debug("Original name: %s." % original)
log.debug("TVDB ID: %s." % tvdb_id)
log.debug("Season: %s episode: %s." % (season, episode))
log.debug("Sonarr series ID: %s." % seriesid)

try:
    try:
        inputfile = renameFile(inputfile, log)
    except:
        log.exception("Error renaming inputfile")

    success = mp.fullprocess(inputfile, MediaType.TV, tvdbid=tvdb_id, season=season, episode=episode, original=original)

    if success:
        # Update Sonarr to continue monitored status
        try:
            host = settings.Sonarr['host']
            port = settings.Sonarr['port']
            webroot = settings.Sonarr['webroot']
            apikey = settings.Sonarr['apikey']
            ssl = settings.Sonarr['ssl']
            protocol = "https://" if ssl else "http://"

            log.debug("Sonarr host: %s." % host)
            log.debug("Sonarr port: %s." % port)
            log.debug("Sonarr webroot: %s." % webroot)
            log.debug("Sonarr apikey: %s." % apikey)
            log.debug("Sonarr protocol: %s." % protocol)

            if apikey != '':
                headers = {'X-Api-Key': apikey}

                if rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log):
                    log.info("Rescan command completed")

                    sonarrepinfo = getEpisodeInformation(host, port, webroot, apikey, protocol, seriesid, log)
                    if not sonarrepinfo:
                        log.error("No valid episode information found, aborting.")
                        sys.exit(1)

                    if not sonarrepinfo.get('hasFile'):
                        log.warning("Rescanned episode does not have a file, attempting second rescan.")
                        if rescanAndWait(host, port, webroot, apikey, protocol, seriesid, log):
                            sonarrepinfo = getEpisodeInformation(host, port, webroot, apikey, protocol, seriesid, log)
                            if not sonarrepinfo:
                                log.error("No valid episode information found, aborting.")
                                sys.exit(1)
                            if not sonarrepinfo.get('hasFile'):
                                log.warning("Rescanned episode still does not have a file, will not set to monitored to prevent endless loop.")
                                sys.exit(1)
                            else:
                                log.info("File found after second rescan.")
                        else:
                            log.error("Rescan command timed out")
                            sys.exit(1)

                    renameSeries(host, port, webroot, apikey, protocol, seriesid, log)

                    # Then set that episode to monitored
                    sonarrepinfo['monitored'] = True
                    log.debug("Sending PUT request with following payload:")
                    log.debug(str(sonarrepinfo))

                    url = protocol + host + ":" + str(port) + webroot + "/api/episode/" + str(sonarrepinfo['id'])
                    r = requests.put(url, json=sonarrepinfo, headers=headers)
                    success = r.json()

                    log.debug("PUT request returned:")
                    log.debug(str(success))
                    log.info("Sonarr monitoring information updated for episode %s." % success['title'])
                else:
                    log.error("Rescan command timed out")
                    sys.exit(1)
            else:
                log.error("Your Sonarr API Key is blank. Update autoProcess.ini to enable status updates.")
        except:
            log.exception("Sonarr monitor status update failed.")
    else:
        log.info("Processing returned False.")
        sys.exit(1)
except:
    log.exception("Error processing file")
    sys.exit(1)
```

I moved the renaming process ahead of setting the episode as 'monitored'

**Comment by joebeem** [comment]
Thank you! I appreciate all the work you do. I am going to try it tonight and will report back.

**Comment by mdhiggins** [comment]
Nah I haven't updated the radarr script but this should be fine. I'm not 100% confident this will work since I think the rename will ignore the file if it hasn't been scanned yet but I'm not sure

The logs all look like its working correctly and renaming the files appropriately so not sure why you get stuck in this infinite loop

I also just pushed an update that makes this new force rename trick optional which you can see here #1325 so you can probably just turn this off and not worry about it

**Comment by mdhiggins** [comment]
@yekul Glad the renaming feature was working for you but it was causing issues for some workflows for other users and so I've made it an optional feature. The rename option is defaulted to `False` so you'll need to turn on the `force-rename` feature in the Sonarr/Radarr settings of autoProcess.ini to have that functionality now, just as a heads up in case you auto update

**Comment by joebeem** [comment]
@mdhiggins Thank you for your time and efforts with my issue and for the whole project..
